src/core/gui/inputdevices/GeometryToolInputHandler.cpp

```diff
@@ -0,0 +1,335 @@
+
+        double xdir = 0;
+        double ydir = 0;
+        double angle = 0.0;
+        double scale = 1.0;
+
+        switch (keyEvent->keyval) {
+            case GDK_KEY_Left:
+                xdir = -1;
+                break;
+            case GDK_KEY_Up:
+                ydir = -1;
+                break;
+            case GDK_KEY_Right:
+                xdir = 1;
+                break;
+            case GDK_KEY_Down:
+                ydir = 1;
+                break;
+            case GDK_KEY_r:
+            case GDK_KEY_R: {  // r like "rotate"
+                angle = (event.state & GDK_MOD1_MASK) ? ROTATE_AMOUNT_SMALL : ROTATE_AMOUNT;
+                angle = (event.state & GDK_SHIFT_MASK) ? angle : -angle;
+                break;
+            }
+            case GDK_KEY_s:
+            case GDK_KEY_S: {
+                scale = (event.state & GDK_MOD1_MASK) ? SCALE_AMOUNT_SMALL : SCALE_AMOUNT;
+                scale = (event.state & GDK_SHIFT_MASK) ? 1.0 / scale : scale;
+                const auto h = height * scale;
+                if (h > getMaxHeight() || h < getMinHeight()) {
+                    scale = 1.0;
+                }
+                break;
+            }
+            case GDK_KEY_m: {
+                controller->markPoint(translationX, translationY);
+                return true;
+            }
+        }
+
+        if (xdir != 0 || ydir != 0) {
+            const auto c = std::cos(rotation);
+            const auto s = std::sin(rotation);
+            double xshift{0.0};
+            double yshift{0.0};
+            const auto amount = (event.state & GDK_MOD1_MASK) ? MOVE_AMOUNT_SMALL : MOVE_AMOUNT;
+            if (event.state & GDK_SHIFT_MASK) {
+                xshift = amount * (c * xdir - s * ydir);
+                yshift = amount * (s * xdir + c * ydir);
+            } else {
+                xshift = amount * xdir;
+                yshift = amount * ydir;
+            }
+            controller->move(xshift, yshift);
+            return true;
+        }
+
+        auto p = utl::Point(translationX, translationY);
+        if (angle != 0) {
+            controller->rotate(angle, p.x, p.y);
+            return true;
+        }
+        if (scale != 1.0) {
+            controller->scale(scale, p.x, p.y);
+            return true;
+        }
+    }
+    return false;
+}
+
+void GeometryToolInputHandler::sequenceStart(InputEvent const& event) {
+    if (event.sequence == this->primarySequence) {
+        this->priLastPageRel = this->getCoords(event);
+    } else {
+        this->secLastPageRel = this->getCoords(event);
+    }
+    const auto page = controller->getPage();
+
+    const auto l = page->getSelectedLayer();
+    this->lines.clear();
+    for (const auto& e: l->getElements()) {
+        if (e->getType() == ELEMENT_STROKE) {
+            auto s = dynamic_cast<Stroke*>(e);
+            if (s->getPointCount() == 2) {
+                lines.push_back(s);
+            }
+        }
+    }
```

Comment:
I find this weird (even if the above comment about using `sequenceStart()` is resolved).

Filtering the entire list of elements everytime the user puts a finger on the screen is bad, performance-wise. I'm not sure how to handle the issue.

[ Also, for any filtering like that, the document's mutex needs to be locked: if an element is added to the layer by another thread, all the iterators could be invalidated, leading to a segfault ]

Reply:
I added the document locking. 
Performance-wise I think there would have to be million or billions of elements on the layer to notice any difference. Under such circumstances other operations (like erasing) might become the bigger problem.
Whilst it would be possible to cache these lines and update the cache whenever needed, I think that would be more error prone and maybe not even be more performant.
Also note that users don't put a new finger on the screen every millisecond or so.